### ISSUE TITLE
Array module review followup: Move dim and dims into array type

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18085]
This is a followup item for the [Array module review](https://github.com/Cray/chapel-private/issues/2241).

The domain datatype has functions:

```
  proc dim(d: int)
  proc dims()
```

Given that the array type already has several functions that effectively forward to the underlying domain it makes sense to add these two in as well.

**Comment by bradcray** [comment]
I don't disagree with this proposal per se, but can you remind me:  Did we come up with any sort of heuristic to decide when to provide these convenience methods vs. when to require the user to refer to the domain to get at information?

**Comment by stonea** [comment]
> I don't disagree with this proposal per se, but can you remind me: Did we come up with any sort of heuristic to decide when to provide these convenience methods vs. when to require the user to refer to the domain to get at information?

No, I don't think we came up with a well defined heuristic\philosophy in the meeting to decide what should go where.

Maybe the heuristic should be that methods that one would typically find on a rectangular array datatype from other languages should also be on the array datatype in Chapel. 

This first half of this comment from @mppf (in our meeting notes) echoes this, although I think maybe he's advocating for the opposite in the second half (not sure if the word "that" in the last sentence is refers to operating on arrays and not knowing about domains or whether it refers to leaning more on domains than on arrays):

> Michael: people are used to operating on arrays and not knowing about domains.
Having this stuff there is convenient for people that aren't used to it.
Eventually they should learn to lean more on domains than arrays, but it's not
bad that they do from the start.

Anyway I put this issue in my "uncontroversial changes" bucket as we took a vote on it during the meeting:

> Andy: are we decided that we can move dim/dims into array?  Anyone against it?
(No objections voiced)

**Comment by mppf** [comment]
Yeah, you can pretty much ignore the part after `that` in my remark. I think we should have these convenience functions on arrays.

**Comment by bradcray** [comment]
That rationale works for me, thanks @stonea (I couldn't remember where this fell on the controversial vs. not scale).
